test/jdk/java/lang/ref/FinalizerHistogramTest.java

```diff
@@ -32,37 +33,50 @@
  * @test
  * @summary Unit test for FinalizerHistogram
  * @modules java.base/java.lang.ref:open
- * @run main FinalizerHistogramTest
+ * @library /test/lib
+ * @build jdk.test.whitebox.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
+ * @run main/othervm
+ *      -Xbootclasspath/a:.
+ *      -XX:+UnlockDiagnosticVMOptions
+ *      -XX:+WhiteBoxAPI
+ *      FinalizerHistogramTest
  */
 
 public class FinalizerHistogramTest {
     static ReentrantLock lock = new ReentrantLock();
-    static volatile int wasInitialized = 0;
-    static volatile int wasTrapped = 0;
-    static final int objectsCount = 1000;
+    static final AtomicInteger initializedCount = new AtomicInteger(0);
+    static final AtomicInteger trappedCount = new AtomicInteger(0);
+    static final int OBJECTS_COUNT = 1000;
+
+    static WhiteBox wb;
+    static boolean refProResult;
 
     static class MyObject {
         public MyObject() {
             // Make sure object allocation/deallocation is not optimized out
-            wasInitialized += 1;
+            initializedCount.incrementAndGet();
         }
 
         protected void finalize() {
             // Trap the object in a finalization queue
-            wasTrapped += 1;
+            trappedCount.incrementAndGet();
             lock.lock();
         }
     }
 
-    public static void main(String[] argvs) {
+    public static void main(String[] argvs) throws InterruptedException {
         try {
             lock.lock();
-            for(int i = 0; i < objectsCount; ++i) {
+            for(int i = 0; i < OBJECTS_COUNT; ++i) {
                 new MyObject();
             }
-            System.out.println("Objects intialized: " + objectsCount);
-            System.gc();
-            while(wasTrapped < 1);
+            System.out.println("Objects intialized: " + initializedCount.get());
+            wb = WhiteBox.getWhiteBox();
+            wb.fullGC();
+            refProResult = wb.waitForReferenceProcessing();
+            System.out.println("waitForReferenceProcessing returned: " + refProResult);
+            while(trappedCount.get() < 1);
```

Comment:
This doesn't seem to be doing anything?  trappedCount is initially 0, and only incremented.

Reply:
> It spins until at least one `MyObject` has entered `finalize()` before proceeding to fetch the histogram.

My eyes must be going.  Somehow I read that as a `-1`.

Reply:
I still think this loop isn't actually needed.  What the test is interested in is having final references in the
finalizer queue.  Waiting for reference processing does that.  Waiting for trappedCount to be non-zero
(really, waiting for 1, with more than one indicating a problem with the locking and possible failure to
keep the objects in the finalizer queue), isn't necessary for the remainder of the test to work.  I think
trappedCount isn't needed at all in the current version, and should either be removed or made useful in
some fashion.